File: algorithms/searching/knuth-morris-pratt/implementations/java/KMPSearch.java
/**
 * Knuth-Morris-Pratt (KMP) String Matching Algorithm Implementation in Java
 * KMP字符串匹配算法的Java实现
 * 
 * Time Complexity: O(n + m) where n is text length and m is pattern length
 * Space Complexity: O(m) for the failure function array
 * 
 * @author Algorithm Collection Repository
 * @version 1.0
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class KMPSearch {
    
    /**
     * Build the failure function (LPS array) for the KMP algorithm.
     * 
     * The failure function stores the length of the longest proper prefix 
     * which is also a suffix for each position in the pattern.
     * 
     * @param pattern The pattern string for which to build the failure function
     * @return Failure function array where lps[i] represents the length of 
     *         the longest proper prefix which is also a suffix for pattern[0:i+1]
     * @throws IllegalArgumentException if pattern is null
     */
    public static int[] buildFailureFunction(String pattern) {
        if (pattern == null) {
            throw new IllegalArgumentException("Pattern cannot be null");
        }
        
        if (pattern.isEmpty()) {
            return new int[0];
        }
        
        int m = pattern.length();
        int[] lps = new int[m]; // Longest Proper Prefix which is also Suffix
        int length = 0; // Length of the previous longest prefix suffix
        int i = 1;      // Start from the second character
        
        // Build the failure function
        while (i < m) {
            if (pattern.charAt(i) == pattern.charAt(length)) {
                length++;
                lps[i] = length;
                i++;
            } else {
                // This is tricky. Consider the example "AAACAAAA" and i = 7.
                // The idea is similar to search algorithm.
                if (length != 0) {
                    length = lps[length - 1];
                    // Also, note that we do not increment i here
                } else {
                    lps[i] = 0;
                    i++;
                }
            }
        }
        
        return lps;
    }
    
    /**
     * Search for all occurrences of pattern in text using KMP algorithm.
     * 
     * @param text The text string in which to search
     * @param pattern The pattern string to search for
     * @return List of starting positions where pattern is found in text
     * @throws IllegalArgumentException if text or pattern is null
     */
    public static List<Integer> search(String text, String pattern) {
        if (text == null) {
            throw new IllegalArgumentException("Text cannot be null");
        }
        if (pattern == null) {
            throw new IllegalArgumentException("Pattern cannot be null");
        }
        
        List<Integer> positions = new ArrayList<>();
        
        if (text.isEmpty() || pattern.isEmpty()) {
            return positions;
        }
        
        int n = text.length();
        int m = pattern.length();
        
        // Get the failure function for the pattern
        int[] lps = buildFailureFunction(pattern);
        
        int i = 0; // Index for text
        int j = 0; // Index for pattern
        
        while (i < n) {
            if (pattern.charAt(j) == text.charAt(i)) {
                j++;
                i++;
            }
            
            if (j == m) {
                // Found a match
                positions.add(i - j);
                j = lps[j - 1];
            } else if (i < n && pattern.charAt(j) != text.charAt(i)) {
                // Do not match j with lps[0..lps[j-1]] characters,
                // they will match anyway
                if (j != 0) {
                    j = lps[j - 1];
                } else {
                    i++;
                }
            }
        }
        
        return positions;
    }
    
    /**
     * Search for the first occurrence of pattern in text using KMP algorithm.
     * 
     * @param text The text string in which to search
     * @param pattern The pattern string to search for
     * @return Starting position of first match, or -1 if not found
     * @throws IllegalArgumentException if text or pattern is null
     */
    public static int searchFirst(String text, String pattern) {
        List<Integer> positions = search(text, pattern);
        return positions.isEmpty() ? -1 : positions.get(0);
    }
    
    /**
     * Validate input parameters for KMP search functions.
     * 
     * @param text The text string
     * @param pattern The pattern string
     * @throws IllegalArgumentException if inputs are null or pattern is empty
     */
    public static void validateInput(String text, String pattern) {
        if (text == null) {
            throw new IllegalArgumentException("Text cannot be null");
        }
        if (pattern == null) {
            throw new IllegalArgumentException("Pattern cannot be null");
        }
        if (pattern.isEmpty()) {
            throw new IllegalArgumentException("Pattern cannot be empty");
        }
    }
    
    /**
     * A matcher class that allows for pattern preprocessing and multiple searches
     * with the same pattern for better performance.
     */
    public static class KMPMatcher {
        private final String pattern;
        private final int[] failureFunction;
        
        /**
         * Initialize the KMP matcher with a pattern.
         * 
         * @param pattern The pattern to search for
         * @throws IllegalArgumentException if pattern is null or empty
         */
        public KMPMatcher(String pattern) {
            if (pattern == null) {
                throw new IllegalArgumentException("Pattern cannot be null");
            }
            if (pattern.isEmpty()) {
                throw new IllegalArgumentException("Pattern cannot be empty");
            }
            
            this.pattern = pattern;
            this.failureFunction = buildFailureFunction(pattern);
        }
        
        /**
         * Search for all occurrences of the pattern in the given text.
         * 
         * @param text The text to search in
         * @return List of starting positions where pattern is found
         * @throws IllegalArgumentException if text is null
         */
        public List<Integer> search(String text) {
            if (text == null) {
                throw new IllegalArgumentException("Text cannot be null");
            }
            
            return KMPSearch.search(text, this.pattern);
        }
        
        /**
         * Search for the first occurrence of the pattern in the given text.
         * 
         * @param text The text to search in
         * @return Starting position of first match, or -1 if not found
         * @throws IllegalArgumentException if text is null
         */
        public int searchFirst(String text) {
            List<Integer> positions = search(text);
            return positions.isEmpty() ? -1 : positions.get(0);
        }
        
        /**
         * Get the pattern string.
         * 
         * @return The pattern string
         */
        public String getPattern() {
            return pattern;
        }
        
        /**
         * Get a copy of the failure function (LPS array) for the pattern.
         * 
         * @return Copy of the failure function array
         */
        public int[] getFailureFunction() {
            return Arrays.copyOf(failureFunction, failureFunction.length);
        }
    }
    
    /**
     * Utility method to print array nicely.
     * 
     * @param array The array to print
     * @return String representation of the array
     */
    private static String arrayToString(int[] array) {
        return Arrays.toString(array);
    }
    
    /**
     * Utility method to print list nicely.
     * 
     * @param list The list to print
     * @return String representation of the list
     */
    private static String listToString(List<Integer> list) {
        return list.toString();
    }
    
    /**
     * Example usage and demonstration of the KMP algorithm.
     * 
     * @param args Command line arguments (not used)
     */
    public static void main(String[] args) {
        System.out.println("KMP String Matching Algorithm Demo");
        System.out.println("=".repeat(40));
        
        // Test case 1: Basic search
        String text1 = "ABABDABACDABABCABCABCABCABC";
        String pattern1 = "ABABCAB";
        List<Integer> result1 = search(text1, pattern1);
        System.out.println("Text: " + text1);
        System.out.println("Pattern: " + pattern1);
        System.out.println("Matches found at positions: " + listToString(result1));
        System.out.println();
        
        // Test case 2: Multiple matches
        String text2 = "AABAACAADAABAABA";
        String pattern2 = "AABA";
        List<Integer> result2 = search(text2, pattern2);
        System.out.println("Text: " + text2);
        System.out.println("Pattern: " + pattern2);
        System.out.println("Matches found at positions: " + listToString(result2));
        System.out.println();
        
        // Test case 3: Using KMPMatcher class
        KMPMatcher matcher = new KMPMatcher("ana");
        String text3 = "banana";
        List<Integer> result3 = matcher.search(text3);
        System.out.println("Using KMPMatcher class:");
        System.out.println("Text: " + text3);
        System.out.println("Pattern: " + matcher.getPattern());
        System.out.println("Matches found at positions: " + listToString(result3));
        System.out.println("Failure function: " + arrayToString(matcher.getFailureFunction()));
        System.out.println();
        
        // Test case 4: Failure function demonstration
        String[] patterns = {"ABABCABAB", "AAAA", "ABCDABCD"};
        System.out.println("Failure Function Examples:");
        for (String pattern : patterns) {
            int[] lps = buildFailureFunction(pattern);
            System.out.println("Pattern: " + pattern + " → LPS: " + arrayToString(lps));
        }
        System.out.println();
        
        // Test case 5: Performance with large text
        StringBuilder largeTextBuilder = new StringBuilder();
        for (int i = 0; i < 1000; i++) {
            largeTextBuilder.append("A");
        }
        largeTextBuilder.append("PATTERN");
        for (int i = 0; i < 1000; i++) {
            largeTextBuilder.append("B");
        }
        String largeText = largeTextBuilder.toString();
        
        long startTime = System.nanoTime();
        List<Integer> largeResult = search(largeText, "PATTERN");
        long endTime = System.nanoTime();
        
        System.out.println("Performance Test:");
        System.out.println("Text length: " + largeText.length());
        System.out.println("Pattern found at: " + listToString(largeResult));
        System.out.println("Time taken: " + ((endTime - startTime) / 1_000_000.0) + " ms");
        
        // Test edge cases
        System.out.println("\\nEdge Cases:");
        try {
            validateInput("text", "pattern");
            System.out.println("✓ Valid input validation passed");
        } catch (IllegalArgumentException e) {
            System.out.println("✗ Valid input validation failed: " + e.getMessage());
        }
        
        try {
            search("text", "");
            System.out.println("✓ Empty pattern handled gracefully");
        } catch (IllegalArgumentException e) {
            System.out.println("✓ Empty pattern correctly rejected: " + e.getMessage());
        }
        
        try {
            search(null, "pattern");
            System.out.println("✗ Null text should throw exception");
        } catch (IllegalArgumentException e) {
            System.out.println("✓ Null text correctly rejected: " + e.getMessage());
        }
    }
}

File: algorithms/searching/knuth-morris-pratt/implementations/python/kmp_search.py
"""
Knuth-Morris-Pratt (KMP) String Matching Algorithm Implementation in Python
KMP字符串匹配算法的Python实现

Time Complexity: O(n + m) where n is text length and m is pattern length
Space Complexity: O(m) for the failure function array
"""

def build_failure_function(pattern):
    """
    Build the failure function (LPS array) for the KMP algorithm.
    
    The failure function stores the length of the longest proper prefix 
    which is also a suffix for each position in the pattern.
    
    Args:
        pattern (str): The pattern string for which to build the failure function
        
    Returns:
        list: Failure function array where lps[i] represents the length of 
              the longest proper prefix which is also a suffix for pattern[0:i+1]
              
    Examples:
        >>> build_failure_function("ABABCABAB")
        [0, 0, 1, 2, 0, 1, 2, 1, 2]
        
        >>> build_failure_function("AAAA")
        [0, 1, 2, 3]
        
        >>> build_failure_function("ABCDABCD")
        [0, 0, 0, 0, 1, 2, 3, 4]
        
        >>> build_failure_function("")
        []
    """
    if not pattern:
        return []
    
    m = len(pattern)
    lps = [0] * m  # Longest Proper Prefix which is also Suffix
    length = 0     # Length of the previous longest prefix suffix
    i = 1          # Start from the second character
    
    # Build the failure function
    while i < m:
        if pattern[i] == pattern[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            # This is tricky. Consider the example "AAACAAAA" and i = 7.
            # The idea is similar to search algorithm.
            if length != 0:
                length = lps[length - 1]
                # Also, note that we do not increment i here
            else:
                lps[i] = 0
                i += 1
    
    return lps


def kmp_search(text, pattern):
    """
    Search for all occurrences of pattern in text using KMP algorithm.
    
    Args:
        text (str): The text string in which to search
        pattern (str): The pattern string to search for
        
    Returns:
        list: List of starting positions where pattern is found in text
        
    Examples:
        >>> kmp_search("ABABDABACDABABCABCABCABCABC", "ABABCAB")
        [15]
        
        >>> kmp_search("AABAACAADAABAABA", "AABA")
        [0, 9, 12]
        
        >>> kmp_search("hello world", "world")
        [6]
        
        >>> kmp_search("hello world", "xyz")
        []
        
        >>> kmp_search("", "pattern")
        []
        
        >>> kmp_search("text", "")
        []
    """
    if not text or not pattern:
        return []
    
    n = len(text)
    m = len(pattern)
    
    # Get the failure function for the pattern
    lps = build_failure_function(pattern)
    
    positions = []  # Store all match positions
    i = 0  # Index for text
    j = 0  # Index for pattern
    
    while i < n:
        if pattern[j] == text[i]:
            j += 1
            i += 1
        
        if j == m:
            # Found a match
            positions.append(i - j)
            j = lps[j - 1]
        elif i < n and pattern[j] != text[i]:
            # Do not match j with lps[0..lps[j-1]] characters,
            # they will match anyway
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1
    
    return positions


def kmp_search_first(text, pattern):
    """
    Search for the first occurrence of pattern in text using KMP algorithm.
    
    Args:
        text (str): The text string in which to search
        pattern (str): The pattern string to search for
        
    Returns:
        int: Starting position of first match, or -1 if not found
        
    Examples:
        >>> kmp_search_first("ABABDABACDABABCABCABCABCABC", "ABABCAB")
        15
        
        >>> kmp_search_first("hello world", "world")
        6
        
        >>> kmp_search_first("hello world", "xyz")
        -1
        
        >>> kmp_search_first("", "pattern")
        -1
    """
    positions = kmp_search(text, pattern)
    return positions[0] if positions else -1


def validate_input(text, pattern):
    """
    Validate input parameters for KMP search functions.
    
    Args:
        text (str): The text string
        pattern (str): The pattern string
        
    Raises:
        TypeError: If inputs are not strings
        ValueError: If pattern is empty
        
    Examples:
        >>> validate_input("hello", "lo")
        True
        
        >>> validate_input(123, "pattern")
        Traceback (most recent call last):
        ...
        TypeError: Text must be a string
        
        >>> validate_input("text", "")
        Traceback (most recent call last):
        ...
        ValueError: Pattern cannot be empty
    """
    if not isinstance(text, str):
        raise TypeError("Text must be a string")
    if not isinstance(pattern, str):
        raise TypeError("Pattern must be a string")
    if not pattern:
        raise ValueError("Pattern cannot be empty")
    return True


class KMPMatcher:
    """
    A class-based implementation of KMP string matching algorithm.
    
    This class allows for pattern preprocessing and multiple searches
    with the same pattern for better performance.
    """
    
    def __init__(self, pattern):
        """
        Initialize the KMP matcher with a pattern.
        
        Args:
            pattern (str): The pattern to search for
            
        Examples:
            >>> matcher = KMPMatcher("ABABCAB")
            >>> matcher.pattern
            'ABABCAB'
        """
        if not isinstance(pattern, str):
            raise TypeError("Pattern must be a string")
        if not pattern:
            raise ValueError("Pattern cannot be empty")
            
        self.pattern = pattern
        self.failure_function = build_failure_function(pattern)
    
    def search(self, text):
        """
        Search for all occurrences of the pattern in the given text.
        
        Args:
            text (str): The text to search in
            
        Returns:
            list: List of starting positions where pattern is found
            
        Examples:
            >>> matcher = KMPMatcher("AB")
            >>> matcher.search("ABABAB")
            [0, 2, 4]
        """
        if not isinstance(text, str):
            raise TypeError("Text must be a string")
        
        return kmp_search(text, self.pattern)
    
    def search_first(self, text):
        """
        Search for the first occurrence of the pattern in the given text.
        
        Args:
            text (str): The text to search in
            
        Returns:
            int: Starting position of first match, or -1 if not found
            
        Examples:
            >>> matcher = KMPMatcher("world")
            >>> matcher.search_first("hello world")
            6
        """
        positions = self.search(text)
        return positions[0] if positions else -1


# Example usage and testing
if __name__ == "__main__":
    import doctest
    doctest.testmod()
    
    # Additional manual testing
    print("KMP String Matching Algorithm Demo")
    print("=" * 40)
    
    # Test case 1
    text1 = "ABABDABACDABABCABCABCABCABC"
    pattern1 = "ABABCAB"
    result1 = kmp_search(text1, pattern1)
    print(f"Text: {text1}")
    print(f"Pattern: {pattern1}")
    print(f"Matches found at positions: {result1}")
    print()
    
    # Test case 2
    text2 = "The quick brown fox jumps over the lazy dog"
    pattern2 = "the"
    result2 = kmp_search(text2.lower(), pattern2)
    print(f"Text: {text2}")
    print(f"Pattern: {pattern2}")
    print(f"Matches found at positions: {result2}")
    print()
    
    # Test case 3 - Using KMPMatcher class
    matcher = KMPMatcher("ana")
    text3 = "banana"
    result3 = matcher.search(text3)
    print(f"Using KMPMatcher class:")
    print(f"Text: {text3}")
    print(f"Pattern: {matcher.pattern}")
    print(f"Matches found at positions: {result3}")
    print(f"Failure function: {matcher.failure_function}")